### PR TITLE
Track project dependencies for intermediate model requests

### PR DESCRIPTION
**Pull request by alllex**
When a top-level model is aggregated from intermediate models of other projects, the top-level model must be invalidated if the configuration of any of these projects changes in the case of an incremental reconfiguration.

A specific example is the IP-safe model building for the `GradleProject` that is done through fetching isolated models for individual projects. Another example is the `IdeaProject` model that aggregates all projects into individual `IdeaModule` models.

### CODE REVIEW

File: platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiGradleProjectIntegrationTest.groovy [hunk 1]
```diff
@@ -201,4 +201,59 @@ class IsolatedProjectsToolingApiGradleProjectIntegrationTest extends AbstractIso
         fixture.assertStateLoaded()
     }
 
+    def "root GradleProject model is invalidated when a child project configuration changes"() {
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiGradleProjectIntegrationTest.groovy:204]: 💡 Ideally, the root GradleProject model would only be invalidated when a consumed child project model changes. That sort of dependency tracking is what we expect to have with _Shared Data_ so perhaps some of these models would be migrated to shared data once the time is ripe.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt [hunk 2]
```diff
@@ -86,7 +82,17 @@ class DefaultBuildTreeModelControllerServices : BuildTreeModelControllerServices
         val configurationCacheLogLevel = if (startParameter.isConfigurationCacheQuiet) LogLevel.INFO else LogLevel.LIFECYCLE
         val modelParameters = if (requirements.isCreatesModel) {
             // When creating a model, disable certain features - only enable configure on demand and configuration cache when isolated projects is enabled
-            BuildModelParameters(parallelProjectExecution, isolatedProjects, isolatedProjects, isolatedProjects, true, isolatedProjects, parallelToolingActions, invalidateCoupledProjects, configurationCacheLogLevel)
+            BuildModelParameters(
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt:85]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt [hunk 3]
```diff
@@ -99,7 +105,17 @@ class DefaultBuildTreeModelControllerServices : BuildTreeModelControllerServices
             when {
                 configurationCache && startParameter.writeDependencyVerifications.isNotEmpty() -> disabledConfigurationCacheBuildModelParameters(StartParameterBuildOptions.DependencyVerificationWriteOption.LONG_OPTION)
                 configurationCache && startParameter.isExportKeys -> disabledConfigurationCacheBuildModelParameters(StartParameterBuildOptions.ExportKeysOption.LONG_OPTION)
-                else -> BuildModelParameters(parallelProjectExecution, configureOnDemand, configurationCache, isolatedProjects, false, false, parallelToolingActions, invalidateCoupledProjects, configurationCacheLogLevel)
+                else -> BuildModelParameters(
+                    parallelProjectExecution,
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt:109]: 👍

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt [hunk 4]
```diff
@@ -125,7 +141,8 @@ class DefaultBuildTreeModelControllerServices : BuildTreeModelControllerServices
         return BuildTreeModelControllerServices.Supplier { registration ->
             registration.add(BuildType::class.java, BuildType.TASKS)
             // Configuration cache is not supported for nested build trees
-            val buildModelParameters = BuildModelParameters(startParameter.isParallelProjectExecutionEnabled, startParameter.isConfigureOnDemand, false, false, true, false, false, false, LogLevel.LIFECYCLE)
+            val buildModelParameters =
+                BuildModelParameters(startParameter.isParallelProjectExecutionEnabled, startParameter.isConfigureOnDemand, false, false, true, false, false, false, LogLevel.LIFECYCLE)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt:145]: 💅 I'd move the constructor arguments to separate lines like in the other call sites

File: subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/IntermediateToolingModelProvider.java [hunk 5]
```diff
@@ -31,31 +31,26 @@
...
 public interface IntermediateToolingModelProvider {
 
     /**
-     * Fetches models of a given type for the given projects.
-     * <p>
-     * Model name to find the underlying builder is derived from the binary name of the {@code modelType}.
+     * Fetches models of a given type for the given projects passing a parameter to the underlying builder.
      */
-    <T> List<T> getModels(List<Project> targets, Class<T> modelType);
+    <T> List<T> getModels(Project requester, List<Project> targets, String modelName, Class<T> modelType, Object modelBuilderParameter);
```
Comment (bamboo) [inline subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/IntermediateToolingModelProvider.java:40]: 👍

**Review by bamboo — APPROVED** [review]
👍

**Comment by alllex** [comment]
@bot-gradle merge

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt (hunk 4 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/DefaultBuildTreeModelControllerServices.kt]: Done

File: subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/DefaultIntermediateToolingModelProvider.java [hunk 6]
```diff
@@ -38,47 +38,41 @@ public class DefaultIntermediateToolingModelProvider implements IntermediateTool
...
         this.actionRunner = actionRunner;
         this.parameterCarrierFactory = parameterCarrierFactory;
+        this.projectDependencyListener = projectDependencyListener;
     }
 
     @Override
-    public <T> List<T> getModels(List<Project> targets, Class<T> modelType) {
-        return getModelsImpl(targets, modelType.getName(), modelType, null);
-    }
+    public <T> List<T> getModels(Project requester, List<Project> targets, String modelName, Class<T> modelType, Object modelBuilderParameter) {
```
Comment (6hundreds) [inline subprojects/core/src/main/java/org/gradle/tooling/provider/model/internal/DefaultIntermediateToolingModelProvider.java:54]: 🤔 Might this `requester` name be replaced by more widely using `consumer`? Like you did in `ToolingModelProjectDependencyListener`

**Pull request merged**